src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs

```diff
@@ -62,63 +62,70 @@ internal abstract partial class AbstractNavigateToSearchService
-            {
-                using var _1 = ArrayBuilder<PatternMatch>.GetInstance(out var nameMatches);
-                using var _2 = ArrayBuilder<PatternMatch>.GetInstance(out var containerMatches);
+            // Prioritize the active documents if we have any.
+            var highPriDocs = priorityDocuments.Where(d => project.ContainsDocument(d.Id)).ToSet();
+            await ProcessDocumentsAsync(searchDocument, patternName, patternContainerOpt, declaredSymbolInfoKindsSet, onResultFound, highPriDocs, cancellationToken).ConfigureAwait(false);
+
+            // Then process non-priority documents.
+            var lowPriDocs = project.Documents.Where(d => !highPriDocs.Contains(d)).ToSet();
+            await ProcessDocumentsAsync(searchDocument, patternName, patternContainerOpt, declaredSymbolInfoKindsSet, onResultFound, lowPriDocs, cancellationToken).ConfigureAwait(false);
+
+            // if the caller is only searching a single doc, and we already covered it above, don't bother computing
+            // source-generator docs.
+            if (searchDocument != null && (highPriDocs.Contains(searchDocument) || lowPriDocs.Contains(searchDocument)))
+                return;
+
+            // Finally, generate and process and source-generated docs.  this may take some time, so we always want to
+            // do this after the other documents.
+            var generatedDocs = await project.GetSourceGeneratedDocumentsAsync(cancellationToken).ConfigureAwait(false);
+            await ProcessDocumentsAsync(searchDocument, patternName, patternContainerOpt, declaredSymbolInfoKindsSet, onResultFound, generatedDocs.ToSet<Document>(), cancellationToken).ConfigureAwait(false);
+        }
 
-                var declaredSymbolInfoKindsSet = new DeclaredSymbolInfoKindSet(kinds);
+        private static async Task ProcessDocumentsAsync(
+            Document searchDocument, string patternName, string patternContainerOpt, DeclaredSymbolInfoKindSet kinds,
+            Func<INavigateToSearchResult, Task> onResultFound, ISet<Document> documents, CancellationToken cancellationToken)
+        {
+            using var _ = ArrayBuilder<Task>.GetInstance(out var tasks);
 
-                await ComputeSearchResultsAsync(
-                    project, priorityDocuments, searchDocument, nameMatcher, containerMatcherOpt,
-                    declaredSymbolInfoKindsSet, nameMatches, containerMatches, onResultFound, cancellationToken).ConfigureAwait(false);
+            foreach (var document in documents)
+            {
+                if (searchDocument != null && searchDocument != document)
+                    continue;
+
+                cancellationToken.ThrowIfCancellationRequested();
+                tasks.Add(Task.Run(() =>
+                    ProcessDocumentAsync(document, patternName, patternContainerOpt, kinds, onResultFound, cancellationToken), cancellationToken));
             }
+
+            await Task.WhenAll(tasks).ConfigureAwait(false);
```

Comment:
Is this wait useful if the results are streaming? What is the benefit of letting all high pri documents finish before starting a low pri document?

I assume a high priority document is one that is open? Is the worry that the influx of low priority documents in a large solution will cause some hi-pri to never finish?

Reply:
The general idea is to use the available cpu resources to search the smallest set of priority documents first. That way results in those are find much sooner and reported to the user. Then, we go search the rest of the solution. 

In a situation where the result is in the priority set, the results can come back support quick instead of maybe minutes later. In the end, we always search everything. This just showed us to give preference on where the cpu resources go.